packages/realm-server/migrations/1728561353454_create-user-billing.js

```diff
@@ -0,0 +1,164 @@
+exports.up = (pgm) => {
+  pgm.createTable('users', {
+    id: {
+      type: 'serial',
+      primaryKey: true,
+    },
+    matrix_user_id: { type: 'varchar', notNull: true, unique: true },
+    stripe_customer_id: { type: 'varchar' },
+    created_at: {
+      type: 'timestamp',
+      notNull: true,
+      default: pgm.func('current_timestamp'),
+    },
+  });
+
+  pgm.createTable('plans', {
+    id: {
+      type: 'serial',
+      primaryKey: true,
+    },
+    name: { type: 'varchar', notNull: true },
+    monthly_price: { type: 'numeric(19,4)', notNull: true },
+    credits_included: { type: 'integer', notNull: true },
+    overage_fee_per_credit_in_usd: { type: 'numeric(19,4)' },
+    created_at: {
+      type: 'timestamp',
+      notNull: true,
+      default: pgm.func('current_timestamp'),
+    },
+  });
+
+  pgm.createTable('billing_cycles', {
+    id: {
+      type: 'serial',
+      primaryKey: true,
+    },
+    subscription_id: { type: 'integer' },
+    period_start: { type: 'timestamp', notNull: true },
+    period_end: { type: 'timestamp', notNull: true },
+  });
+
+  pgm.createTable('subscriptions', {
+    id: {
+      type: 'serial',
+      primaryKey: true,
+    },
+    user_id: { type: 'integer', references: 'users(id)' },
+    plan_id: { type: 'integer', references: 'plans(id)' },
+    started_at: { type: 'timestamp', notNull: true },
+    ended_at: { type: 'timestamp', notNull: true },
+    status: { type: 'varchar', notNull: true },
+    stripe_subscription_id: { type: 'varchar', notNull: true },
+    current_billing_cycle_id: {
+      type: 'integer',
+      references: 'billing_cycles(id)',
+    },
+  });
+
+  pgm.addConstraint('billing_cycles', 'billing_cycles_subscription_id_fkey', {
+    foreignKeys: {
+      columns: 'subscription_id',
+      references: 'subscriptions(id)',
+    },
+  });
+
+  pgm.createTable('ai_actions', {
+    id: {
+      type: 'serial',
+      primaryKey: true,
+    },
+    action_type: { type: 'varchar', notNull: true },
+    user_id: { type: 'integer', references: 'users(id)' },
+    cost_in_usd: { type: 'numeric(19,4)', notNull: true },
+    credits_used: { type: 'numeric(19,4)', notNull: true },
+    is_overage: { type: 'boolean', notNull: true, default: false },
+    overage_reported_to_stripe: {
```

Comment:
This supports our plan of batching of overages that need to be reported to Stripe and make a better assurance of not forgetting or overcharging customers for overages

Reply:
I imagine this will be a cron job or something similar that does it periodically every day. Not sure if we defined this already. Comments welcome

Reply:
that seems reasonable. we should be able to fashion a type of job for our workers that is scheduled to run at a particular time in the future with some minor adjustments

Reply:
Sounds good. I would also really, really like to see we implement a retry mechanism for these jobs, so that we can handle stripe webhooks in them and reduce chances of unprocessed events

Reply:
Note to self: i just occurred to me it would be good to also save all stripe webhooks (with raw data). I imagine it would be really helpful to debug potential problems where our system would be out of sync with stripe's